### PR TITLE
Feature: rate limit helpers, and other additions

### PR DESCRIPTION
**Pull request by jonnybazookatone**
**tl; dr** added some tools to help understand the rate limit for each end points, requests made to end points, access to stub data, and not using up the limit during testing.
<hr>
Read the updated README.md if you don't want to read all of the PR comment.

This PR tries to address some of the issues highlighted in #43 #44. This may be a little verbose, but I'd like to explain the reasons for each change for the user case:

# Rate limits
We have had many users hitting the rate limits, not knowing what their limits are, or are not aware there are multiple end points (eg., adsabs/adsabs-dev-api/issues/26; also multiple e-mails) - I'm not complaining, just trying to emphasise this is one of the biggest confusions of using the API at the moment.

To alleviate this, I have made it easier to access the rate limits.

```python
import ads
q = ads.SearchQuery(q=star')
p = q.execute()
q.get_ratelimits()
>>> {'limit': '5000', 'remaining': '4999', 'reset': '1459987200'}
```
This has been implemented by storing the last response in the base class of the Query object. The reasons for this were two fold:
  i) to avoid the situation of having two objects that I execute separately, and would see two different rates, as they do not update one another, ie:

```python
p1 = q1.execute()
q2 = q2.execute()
p1.response.get_ratelimits()
p2.response.get_ratelimits()
assert p1 != p2
>>> True
```
I think this leads to confusion.

  ii) The list() operator should also update the rate limit, which currently does not.

Also, if we have the class keep track of the global rates, then we can have a global utility function that gives all the rates for the known end points:
```python
from ads.utils import get_ratelimits
get_ratelimits()
>>> {'/export': {'limit': None, 'remaining': None, 'reset': None},
 '/metrics': {'limit': None, 'remaining': None, 'reset': None},
 '/search': {'limit': '5000', 'remaining': '4899', 'reset': '1459987200'}}
```

# Damage control

Another issue, related to rate limits, is some users not using the client in an efficient manner. Even though the lazy loading feature is useful for a quick look at the data, it can lead to people unintentionally using up the entire rate limit in one go:

```python
import ads
q = ads.SearchQuery(q='star')
for p in q:
    print p.bibtex
>>> @ARTICLE{...
.... (after 100 articles)
429 Rate limit hit
```

A second issue, is people testing on the live service and then requesting their rate limits be increased *while they are testing*. For the more hands on developers, I've added the `mocks.py` to the `__init__.py` for the `ads` module and also in `setup.py`. People can now freely mock responses if they have a favourite package:
```python
from ads.mocks import MockMetricsResponse
```

However, a lot of the users don't care for this, or understand it, and so I've included a *test* flag, that does this for them without them having to understand:
```python
import ads
q = ads.SearchQuery(q='star', test=True)
p = list(q)
print p[0].title
>>> [u'Title from stub data']
```

To do this, the easiest way seemed to place it in `BaseQuery` as a method that wrapped the normal `execute()` method, such that `HTTPretty` is used if it is a *test*. This required the classes having a relevant `MockResponse` property to be given to HTTPretty. Given that the `Article` class can also lazy load, which instantiates a new `BaseQuery` object, the `Article` class needs to keep track of the parent to understand if the query is a test or not.  A global value could have been used, but this way seemed more fluent to me, but requires keeping track of the `Article` objects' parent.

# Optimisation, and proper use of the client

It also seems it's not straight forward to understand why you are getting a 429, if you don't know the underlying API completely. For example, requesting metrics on an article can also lead to you hitting the metrics limit, and thinking you have hit the limit on the search query end point. To make people understand and trouble shoot their queries, I've added a method on the `BaseQuery` object to see how many requests they've made with a given object, and to which end point:
```python
import ads
q = ads.SearchQuery(q='star', fl=['bibcode'])
p = list(q)
q.get_request_stats()
>>> {'/search': {'requests': 1, 'test_requests': 0}}
...
print p[0].metrics
....
p.get_request_stats()
>>> {'/search': {'requests': 1, 'test_requests': 0}
>>>'/metrics': {'requests': 1, 'test_requests': 0}}
```

The test mode will update the `test_requests` key, which is a little overkill on verbosity. This also requires the `Article` object to know it's parent so that it it can update the request count on the relevant instantiated `BaseQuery` object. I didn't do it at the class level, as this would be propagated to all of the instantiated objects, and make it difficult to understand which query object might be the issue of some rate limit problem.

# Bits and bats

I fixed some of the examples to work with the new client API:
  * `top-cited-astronomer.py` #14 #31
  * `beer-for-cites.py`
  * `journals.py`

The `citation` and `reference`  field returned by the search engine shouldn't be used. Instead one should use the operators `citations()` and `references()`. I updated the `cached-property` fields appropriately.

If you read everything, I'll buy you a pint.

### CODE REVIEW

**Comment by vsudilov** [comment]
Hi Jonny,

Awesome work and great insight into usage patterns and issues.

I agree with all of the added functionalty and fixes, and I think the PR as it stands is a clear improvement to the usability of the library.
I think your implementation is a great first iteration, and after looking into the details there are a few ways I can see to improve it:

#### "Test" mode

Passing around a "testing" flag seems a bit unwieldy. Classes have to keep track of data that isn't relevant to their main purpose, and there are now multiple code-paths (in multiple classes) that need to be maintained and tested.

Moving the "test" mode from in-class to above-class would alleviate these issues, and it would require the following:

  1.  Implementing a ratelimit-decrementing header response in the HTTPretty mocks
      * should be possible by using a generator in the HTTPretty `body` callback.
  2. Deciding on an API for `calling HTTPretty.activate()`
      * Tests already do this via the `with MockResponse():` context manager, and we could add that info to the documentation. An alternative approach could be:
          * `ads.config.testing = True` 
            * ads.config is an object whose `testing` property calls `activate()` or `deactivate()` on `True`/`False`

#### Ratelimits

Explicitly tracking the state of each `APIResponse` via the `parent/_parent` construct throughout the inheritance chain is pretty complex. I think we could achieve the same functionality in a simpler way by having a ratelimit class that is the sole interface to that data. We could take the `logging.getLogger(self.__class__.__name__)` approach to create a named singleton for metrics, search, and export. Something like the following could work ? (pseudo-code):

```
class APIResponse(object):
    @classmethod
    def load_http_response(cls, HTTPResponse):
        if not HTTPResponse.ok:
            raise APIResponseError(HTTPResponse.text)
        c = cls(HTTPResponse.text)
        c.response = HTTPResponse
        Ratelimits.getRatelimits(cls.__name__).set(response.headers)
        return c

    def get_ratelimits(self):
      return Ratelimits.getRateLimits(cls.__name__).to_dict()

class Ratelimits(Singleton):
  # superclass should define __new__ correctly

  @classmethod
  def getRateLimits(cls, name):
    return cls(name)

  def __init__(self, name):
    self.name = name
    self.limits = {}

  def set(self, headers):
    self.limits = headers

  def to_dict(self):
    return self.limits


q = SearchQuery(q="*")
q.execute()
Ratelimits("SearchQuery").to_dict() # {'remaining': 4999}
q.response.get_ratelimits() # {'remaining': 4999}

list(q)
Ratelimits("SearchQuery").to_dict() # {'remaining': 4800}
q.response.get_ratelimits() # {'remaining': 4800}

q2 = SearchQuery(q="new query")
q2.execute()
q2.response.get_ratelimits() == q.response.get_ratelimits() # True
```

Additionally, the library shouldn't handle computing ratelimits -- it should only report the most up-to-date info that the ADS-API returned.

You alluded to an interesting point regarding the usage of on-demand field loading via article attributes. Do you think this particular feature causes more harm than good? Would it better serve users to replace this feature with something like `Exception("You should specify {field} in the fl= argument")`?

**Comment by jonnybazookatone** [comment]
Thanks for taking a look and giving feedback, what's your poison?

I agree, keeping track of the testing state within the class can be cumbersome, but does remove some of the more useful (imo) features as a user when keeping the test state within the instantiated object. For example:

```python
q1 = ads.SearchQuery(q='star', test=True)
papers1 = list(q1)
q2 = ads.SearchQuery(q='star')
papers2 = list(q2)
print papers1.citation_count
```
vs.

```python
q1 = ads.SearchQuery(q='star')
ads.config.testing = True
papers1 = list(q1)
ads.config.testing = False
q2 = ads.SearchQuery(q='star')
papers2 = list(q2)
ads.config.testing = True
papers1.citation_count
```

Normally, I would agree the second is fine, but I can see it being more confusing for some. It would be nice to keep this work flow. However, I don't see a simple way of achieving this given the way `Article` creates a new `Query` object on cached properties, without retaining the users choice of test state.

Can you expand on your statement of 
>Additionally, the library shouldn't handle computing ratelimits -- it should only report the most up-to-date info that the ADS-API returned.

Currently, the query sub-classes (metrics, search, export) are updating the class attribute `_response`, and aren't computing any rate limits within the library itself - ie., the code you gave above just moves `_response` to being managed by your `RateLimit` singleton. Or you don't like the fact you can access the rate limit via the query object `QueryObject.get_ratelimits` vs `QueryObject.response.get_ratelimits`? 

`_parent` was used for keeping track of the `test` state and also updating the counter on number of requests made for a given query object. How would you suggest to keep this work flow also, with its current granularity? We could have a global generic counter, but then this is similar to the `ads.config.testing` case.

Finally, yes the `fl` field and cached properties do seem to be causing some issues for some people. I'm still torn a little, but I think it's better to keep things less restrictive. I like the flexibility it gives while you're playing around in the interpretor - without you requiring to understand everything straight away.

I think the only way to avoid any issues due to `fl` is to supply more information in the docs about how to use the API appropriately, which should remove people's misunderstanding. Plus, if you raise an exception, the person could still just supply minimal fields `fl=['bibcode']` to make it stop complaining, and still run into the issues when accessing `bibtex` or `metrics`.

**Comment by vsudilov** [comment]
1. I don't think a per-instance testing flag is worth the added complexity -- The query objects shouldn't (have to) be aware if its talking to backend B or MockedB. Is there a way to implement per-instance testing flags while maintaining the objects' backend-agnostic nature?

2. I prefer the singleton pattern in this case because it provides a single, global interface to that data. If setting that data on a class attribute is the way you want to go, why not just set it directly on a `SearchQuery.limits` attribute (forgo trying to bubble that info back up to the Base classes)?

3. I was referring to the "increment" methods -- I didn't understand that these are a means to gauge how many requests have been fired as a complimentary information to the ADS reported ratelimits.

**Comment by andycasey** [comment]
Hi @jonnybazookatone:

This is a spectacular contribution! I knew a few people who were regularly hitting the API limit, but I didn't realise it was such a common issue. The rate functionality you've put in is excellent for optimising API usage.

I agree with @vsudilov that the per-instance testing flag seems a bit,.. tricky. What if we had something like `sandbox.py` that has a `SearchQuery` function (subclassed from `ads.search.SearchQuery`), where `execute` is overwritten such that the requests get pushed through to a `MockSolrResponse`? The reason I am wondering if it could be done this way is because the use-case during testing would be something like:

````python

import ads.sandbox as ads

q = ads.SearchQuery(q='star')

...

# lots of testing crap
````

And then once I've finished testing things and I'm sure (1) I won't hit the rate limit, and (2) the code is optimized for the API, I can just replace

````python
import ads.sandbox as ads
````
with
````python
import ads
````

What do you think @jonnybazookatone @vsudilov?

On a related note,.. I feel like I have inadvertently wasted a lot of ADS' resources by encouraging the lazy-loading. I do feel like having it is very useful for the non-regular user, or if you are just actively exploring things in a terminal/notebook style session, but we should probably do more to encourage proper use of the API through `fl` flags. 

Maybe we could have a one-time warning that is raised when the first lazy-loading is done -- or perhaps through an `ads.config` flag to warn on the `N`th lazy-loading, in the same way `matplotlib` warns you once 20 figures have been opened. At the very least it will help make users more aware of the impact of lazy-loading, in the same way @jonnybazookatone 's updated `README.md` does. Thoughts?

**Comment by vsudilov** [comment]
I like `ads.sandbox` :+1: 

I was also considering adding some sort of notification on lazy-load; that seems like a viable way to encourage good usage without stripping down features. If we do that, let's create a separate issue/PR for it.

**Comment by jonnybazookatone** [comment]
Hmmm, I had replied to this a week ago, but it never appeared here it seemed =[.

I also vote for `ads.sandbox`, it gives people less excuse for making mistakes r.e. rate limits.

Adding warnings is OK, but I would vote against raising any exceptions.

Do you want `ads.sandbox` in a separate PR, separated from the rate limit singleton?

**Comment by vsudilov** [comment]
Seems like two orthogonal features -> two PRs. But if it's a hassle to separate it out, don't worry about it.

**Comment by jonnybazookatone** [comment]
All changes linked have made this PR obsolete.
